cachi2/core/models/sbom.py

```diff
@@ -93,6 +106,235 @@ def _unique_components(cls, components: list[Component]) -> list[Component]:
+        return hash((self.referenceLocator, self.referenceType, self.referenceCategory))
+
+
+class SPDXPackageExternalRefSecurity(SPDXPackageExternalRef):
+    """SPDX Package External Reference for category package-manager.
+
+    Compliant to the SPDX specification:
+    https://spdx.github.io/spdx-spec/v2.3/package-information/#721-external-reference-field
+    """
+
+    referenceCategory: Literal["SECURITY"]
+
+
+class SPDXPackageExternalRefPackageManager(SPDXPackageExternalRef):
+    """SPDX Package External Reference for category package-manager.
+
+    Compliant to the SPDX specification:
+    https://spdx.github.io/spdx-spec/v2.3/package-information/#721-external-reference-field
+    """
+
+    referenceCategory: Literal["PACKAGE-MANAGER"]
+
+
+class SPDXPackageExternalRefPackageManagerPURL(
+    SPDXPackageExternalRefPackageManager, SPDXPackageExternalRefReferenceLocatorURI
+):
+    """SPDX Package External Reference for category package-manager and type purl.
+
+    Compliant to the SPDX specification:
+    https://spdx.github.io/spdx-spec/v2.3/package-information/#721-external-reference-field
+    """
+
+    referenceCategory: Literal["PACKAGE-MANAGER"]
+    referenceType: Literal["purl"]
+
+
+class SPDXPackageExternalRefSecurityPURL(
+    SPDXPackageExternalRefSecurity, SPDXPackageExternalRefReferenceLocatorURI
+):
+    """SPDX Package External Reference for category package-manager and type purl.
+
+    Compliant to the SPDX specification:
+    https://spdx.github.io/spdx-spec/v2.3/package-information/#721-external-reference-field
+    """
+
+    referenceCategory: Literal["SECURITY"]
+    referenceType: Literal["cpe23Type"]
+
+
+SPDXPackageExternalRefPackageManagerType = Annotated[
+    SPDXPackageExternalRefPackageManagerPURL,
+    pydantic.Field(discriminator="referenceType"),
+]
+
+SPDXPackageExternalRefSecurityType = Annotated[
+    SPDXPackageExternalRefSecurityPURL,
+    pydantic.Field(discriminator="referenceType"),
+]
+
+
+SPDXPackageExternalRefType = Annotated[
+    Union[SPDXPackageExternalRefPackageManagerType, SPDXPackageExternalRefSecurityType],
+    pydantic.Field(discriminator="referenceCategory"),
+]
+
+
+class SPDXPackageAnnotation(pydantic.BaseModel):
+    """SPDX Package Annotation.
+
+    Compliant to the SPDX specification:
+    https://github.com/spdx/spdx-spec/blob/development/v2.3/schemas/spdx-schema.json#L237
+    """
+
+    model_config = pydantic.ConfigDict(frozen=True)
+
+    annotator: str
+    annotationDate: str
+    annotationType: Literal["OTHER", "REVIEW"]
+    comment: str
+
+    def __hash__(self) -> int:
+        return hash((self.annotator, self.annotationDate, self.annotationType, self.comment))
+
+
+def _extract_purls(from_refs: list[SPDXPackageExternalRefType]) -> list[str]:
+    return [ref.referenceLocator for ref in from_refs if ref.referenceType == "purl"]
+
+
+def _parse_purls(purls: list[str]) -> list[PackageURL]:
+    return [PackageURL.from_string(purl) for purl in purls if purl]
```

Comment:
These 2 seem to only be used from the `SPDXPackage` model, so you might want to just make these instance methods instead.

Reply:
Correct, but that would make invocations significantly longer which, in turn would remove all benefits from having a helper with a short name. I would rather kept them on module level if you don't mind.